Text/Megaparsec/Error.hs

```diff
@@ -248,6 +252,98 @@ parseErrorPretty
 parseErrorPretty e =
   sourcePosStackPretty (errorPos e) <> ":\n" <> parseErrorTextPretty e
 
+
+-- | Pretty-print a 'ParseError' and display the line on which the parse error
+-- occurred. The rendered 'String' always ends with a newline.
+--
+-- Assumes that the stream tokens can be coerced to an @Int@ and that the value
+-- integer value @10@ corresponds to a new-line character.
+--
+-- @since 6.0.0
+
+parseErrorPrettyWithLine
+  :: ( Enum (Token s)
+     , Ord t
+     , ShowToken t
+     , ShowErrorComponent e
+     , Stream s )
+  => s              -- ^ Original input stream
+  -> ParseError t e -- ^ Parse error to render
+  -> String         -- ^ Result of rendering
+parseErrorPrettyWithLine stream e = mconcat messageComponents
+  where
+    position = errorPos e
+    lineMessage = renderLineMessage stream $ NE.last position
+    messageComponents =
+      [ sourcePosStackPretty position
+      , ":\n"
+      , lineMessage 
+      , parseErrorTextPretty e ] 
+
+-- | Renders the source context to improve error message quality.
+
+renderLineMessage
+  :: forall s .
+     ( Enum (Token s)
+     , Stream s )
+  => s -> SourcePos -> String
+renderLineMessage s p = unlines [ paddingLine, contextLine, pointingLine ]
+  where
+    paddingLine   = padding <> "|"
+    contextLine   = " " <> lineNumberStr <> " | " <> cs
+    pointingLine  = padding <> "|" <> replicate columnNum ' ' <> "^"
+
+    lineNumberStr = show $ unPos linePos
+    padLength = length lineNumberStr + 2
+    padding = replicate padLength ' '
+
+    cs =
+      case ts of
+        [] -> "<empty line>"
+        xs -> chr . fromEnum <$> xs
+    ts = chunkToTokens (Proxy :: Proxy s) lineVal
+    lineVal = streamIter grabLine linePos s
+    linePos = sourceLine p
+    columnNum = unPos $ sourceColumn p
+
+-- | Applies a "taking" function @n@ number of times to a 'Stream' and returns
+-- the last result of the "tking" function, discarding previously taken chunks.
+
+streamIter
+  :: ( Enum (Token s)
+     , Stream s)
+  => (s -> (Tokens s, s)) -- ^ A function which takes part of the stream
+  -> Pos                  -- ^ Number of times to apply the function.
+  -> s                    -- ^ Stream to be taken from
+  -> Tokens s             -- ^ The the last segment taken from the stream
+streamIter f p ts = fst $ go (x-1) (f ts)
+  where
+    x = unPos p
+    go 0 val    = val
+    go n (_,xs) = let val = f xs
+                  in go (n-1) val
+
+-- | Takes a line from a stream.
+--
+-- Tokens which are numericaly coercable to the integer value @10@ are considered
+-- new-line characters.
+
+grabLine :: (Enum (Token s), Stream s) => s -> (Tokens s, s)
+grabLine ts =
+  -- remove the newline from the stream
+  case take1_ ts' of
+      Nothing    -> (line, ts')
+      Just (_,v) -> (line, v  )
+  where
+    -- grab the line content
+    (line, ts') = takeWhile_ isNotNewLine ts
+
+    -- Maybe instead we want to corece to an @Int@, then convert to a @Char@,
```

Comment:
No, that'll be most certainly slower, and `defaultAdvance1` is very sensitive to slow-downs because it's called on every token. Comparison with plain number 10 is OK.